walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,74 @@
 
 package com.google.sps;
 
+import java.io.*;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.ArrayList;
+
 
 public final class FindMeetingQuery {
+  // array representing the number of meetings happening during that minute
+  private ArrayList<Integer> meetings;
+
+  public FindMeetingQuery() {
+    meetings = new ArrayList<Integer>();
+    //before processing the events, the number of meetings for each minute is 0
+    meetings.addAll(Collections.nCopies(TimeRange.END_OF_DAY + 2, 0));
+  }
+
+  /** returns a Collection of time ranges when meeting {@code request} can be scheduled in the day of 
+    * events so that all attendees are free */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      // if the meeting lasts more than a day, there is no solution
+      return new ArrayList<TimeRange>();
+    }
+
+    Collection<String> attendees = request.getAttendees();
+
+    for (Event event : events) {
+      // at least one requested attendee is participating in the event
+      if (event.containsRequestedAttendees(attendees)) {
+        // update the number of meetings during the time of event
+        updateNumberOfMeetings(event.getWhen());
+      }
+    }
+
+    return findAvailableTimeRanges((int)request.getDuration());
+  }
+
+  private void updateNumberOfMeetings(TimeRange when) {
```

Comment:
I'm a bit confused on what's going on in this function.

Shouldn't you increment the number of meetings for every single minute between (start, end), rather than just start and end times?

Some comments on what this function is meant to do would help.

It seems like this only increments start times and decrements end times and then you have `findAvailableTimeRanges` have some additional logic to actually calculate the correct numbers. None of this is really obvious in any way and should be extensively documented.

Reply:
To get the number of meetings, `findAvailableTimeRanges `will compute prefix sums, and that's why only the start and end times are changed for each event.

I thought I should explain the details in the PR, but I will add more comments to the code

Reply:
Sorry I completely missed the details in the PR (my bad).

Offtopic: In general, try to include technical explanations as comments in the code, and keep PR descriptions more focused on high level design decisions. The PR description should help the reviewer understand the context of the PR, how do you intend to break down the functionality, potential alternatives considered, etc.

The technical details of what a variable/method does are not only of interest for the PR reviewer, but also to anyone who is reading the codebase later (after this PR gets merged).